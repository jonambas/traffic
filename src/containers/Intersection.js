import React, { Component, createContext } from 'react';
import { initialPattern, pairs, pattern, durations } from '../config';
import _ from 'lodash';

const defaultContext = {
  intersection: {},
  activePairIndex: 0,
  activePatternIndex: 0
};

export const IntersectionContext = createContext(defaultContext);

class IntersectionProvider extends Component {
  state = defaultContext;

  componentDidMount() { 
    this.updateIntersection();
  }

  updateIntersection = (nextPairIndex = 0, nextPatternIndex = 0) => {
    const activePair = pairs[nextPairIndex];
    const nextPattern = pattern[nextPatternIndex];
    const roads = _.flatten(pairs);

    const state = {
      activePairIndex: nextPairIndex,
      activePatternIndex: nextPatternIndex
    };
    
    /**
     * Describes road state, keyed by road name.
     * @example
     *  east: {left: false, straight: true}
     *  north: {left: false, straight: false}
     *  south: {left: false, straight: false}
     *  west: {left: false, straight: true}
     */
    const intersection = roads.reduce((acc, road) => ({
      ...acc,
      [road]: _.includes(activePair, road) ? nextPattern : initialPattern
    }), {});

    this.setState({ ...state, intersection }, this.next);
  }

  /**
   * Increments activePairIndex and activePatternIndex
   */
  next = () => {
    setTimeout(() => {
      const { activePairIndex, activePatternIndex } = this.state;
      const nextPatternIndex = activePatternIndex + 1;
      let nextPairIndex = activePairIndex;
      
      if (nextPatternIndex >= pattern.length) {
        nextPairIndex = activePairIndex + 1;
      }
      
      this.updateIntersection(nextPairIndex % pairs.length, nextPatternIndex % pattern.length);
    }, durations.patternDuration)
  }

  render() {
    console.log(this.state.intersection)

    return (
      <IntersectionContext.Provider value={this.state}>
        {this.props.children}
      </IntersectionContext.Provider>
    )
  }
}

export default IntersectionProvider;