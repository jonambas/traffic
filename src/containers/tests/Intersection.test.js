import React from 'react';
import Intersection from '../Intersection';
import { shallow } from 'enzyme';

describe('Intersection Provider', () => {
  let wrapper;
  let next;

  const roadQueueMock = {
    north: [0,1],
    south: [0,2],
    east: [0,3],
    west: [0,4]
  }

  beforeEach(() => {
    global.setTimeout = jest.fn((fn) => next = fn);
    wrapper = shallow(<Intersection roadQueue={roadQueueMock}>child</Intersection>);
  });

  it('should correctly reshape and set initial state', () => {
    expect(wrapper.prop('value')).toMatchSnapshot();
  });

  it('should correctly step through patterns', async () => {
    for (let i = 0; i < 4; i++) {
      await next();  
      expect(wrapper.prop('value')).toMatchSnapshot(); 
    }
  });
});