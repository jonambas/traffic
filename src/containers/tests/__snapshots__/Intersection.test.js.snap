// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intersection Provider should correctly reshape and set initial state 1`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": true,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": true,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": false,
  },
}
`;

exports[`Intersection Provider should correctly step through patterns 1`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": false,
    "straight": true,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": false,
    "straight": true,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": false,
  },
}
`;

exports[`Intersection Provider should correctly step through patterns 2`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": true,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": false,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": false,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": true,
    "straight": false,
  },
}
`;

exports[`Intersection Provider should correctly step through patterns 3`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": true,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": false,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": false,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": true,
  },
}
`;

exports[`Intersection Provider should correctly step through patterns 4`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": true,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": true,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": false,
  },
}
`;
